openapi3/internalize_refs.go

```diff
@@ -29,6 +33,70 @@ func DefaultRefNameResolver(ref string) string {
 	return filepath.Base(ref)
 }
 
+// UpdatedDefaultRefNameResolver.
+// TODO: docs
+func UpdatedDefaultRefNameResolver(doc *T, ref componentRef) string {
+	if ref.RefString() == "" || ref.RefPath() == nil {
+		return ""
+	}
+
+	refPath := ref.RefPath().String()
+
+	// If refering to a component in the root spec, just use the name
+	if name, found := MatchesComponentInRootDocument(doc, ref); found {
+		c := *ref.RefPath()
+		c.Path = name
+		refPath = c.String()
+	}
+
+	// Cut out the "#/components/schemas" part
+	if b, a, ok := strings.Cut(refPath, path.Join("%23/components/", ref.ComponentType())); ok {
+		refPath = path.Join(b, a)
+	}
+
+	// Trim the common prefix with the root spec path
+	d := doc.url.Path
+	for {
+		if d == "." { // no common prefix
+			break
+		}
+
+		// XXX(alb): The results need clever, respect directory boundaries.
+		if a, found := strings.CutPrefix(refPath, d); found {
+			refPath = a
+			break
+		}
+
+		d = filepath.Dir(d)
+	}
+
+	refPath = strings.TrimLeft(refPath, "/")
+	filePath, componentPath, _ := strings.Cut(refPath, "%23") // refpath is Path encoded
+	refPath = ""
+
+	if filePath != "" {
+		// Remove the path extentions to make it JSON/YAML agnostic.
+		for ext := filepath.Ext(filePath); len(ext) > 0; ext = filepath.Ext(filePath) {
+			filePath = strings.TrimSuffix(filePath, ext)
+		}
+
+		refPath = filePath
+	}
+
+	if componentPath != "" {
+		componentPath = strings.TrimLeft(componentPath, "/")
+		if refPath != "" {
+			refPath += "_"
+		}
+		refPath += componentPath
+	}
+
+	// XXX(alb): These should really be escaped to ~1
+	// But kin-openapi doesn't support refs with ~ at the moment.
```

Comment:
Hm in what way? It's possible I missed some corners

Reply:
If I didn't replace slashes with https://github.com/getkin/kin-openapi/pull/955/files#diff-ae5414b666f1cdb100809e5937dc1a92e1b0010861ec51943ec70faad26019eeR97 I'd get invalid character error and if I encoded the slash to `~1` I'd also get an invalid character error.

I'll expand on this and make another issue if there is one.

Reply:
With either
```
	return strings.ReplaceAll(refPath, "/", "~1") # encode / to ~1
	return refPath # leave slashes as they are
```
I get:
```
 	invalid components: schema "components.openapi.yml/CustomTestSchema": identifier "components.openapi.yml/CustomTestSchema" is not supported by OpenAPIv3 standard (regexp: "^[a-zA-Z0-9._-]+$")
```

Reply:
`/` or `~` are not valid characters in fixed field names which is why I was running into issues in both cases: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#components-object

Opened this issue only to release this shortly after: https://github.com/getkin/kin-openapi/issues/963#issuecomment-2169387253

Therefore swapping to a supported character like `_` might be the solution. This function at least needs to make sure the name it generates only has the following characters: `^[a-zA-Z0-9\.\-_]+$`

Reply:
Just to say I did implement that solution in this PR - any non-supported character is replaced by a `_`.